attachment_interfaces.go

```diff
@@ -0,0 +1,76 @@
+package proton
+
+import (
+	"bytes"
+	"context"
+	"github.com/bradenaw/juniper/parallel"
+)
+
+// AttachmentStorageProvider abstract the attachment download buffer creation.
+type AttachmentStorageProvider interface {
+	// NewBuffer should return a new byte buffer for use. Note that this function may be called from multiple go-routines.
+	NewBuffer() *bytes.Buffer
+}
+
+type DefaultAttachmentStorageProvider struct{}
+
+func NewDefaultAttachmentStorageProvider() *DefaultAttachmentStorageProvider {
+	return &DefaultAttachmentStorageProvider{}
+}
+
+func (DefaultAttachmentStorageProvider) NewBuffer() *bytes.Buffer {
+	return bytes.NewBuffer(nil)
+}
+
+// AttachmentDownloadScheduler allows the user to specify how the attachment data for the message should be downloaded.
+type AttachmentDownloadScheduler interface {
+	Schedule(ctx context.Context, attachmentIDs []string, storageProvider AttachmentStorageProvider, downloader func(string, *bytes.Buffer) error) ([]*bytes.Buffer, error)
+}
+
+// SingleThreadedAttachmentDownloadScheduler downloads the attachments one by one.
+type SingleThreadedAttachmentDownloadScheduler struct{}
```

Comment:
These type names are getting pretty crazy :D 

Given that this allocator/scheduler don't _have_ to be used just for attachments (they would be really useful for proton drive blocks!), why not just name them:
- Allocator
- Scheduler
- SequentialScheduler
- ParallelScheduler

Reply:
Definitely better. Good suggestion! Done.